### DISCUSSION TITLE
New dependency libglpk-dev?

### DISCUSSION THREAD

**Discussion by gorkang** [ropensci/tarchetypes#109] [Help] [answered]
Today when running  a targets pipeline I got this error:

> Error: package or namespace load failed for ‘tarchetypes’ in dyn.load(file, DLLpath = DLLpath, ...):
>   unable to load shared object '/home/USER/R/x86_64-pc-linux-gnu-library/4.2/igraph/libs/igraph.so':
>   libglpk.so.40: cannot open shared object file: No such file or directory

Got rid of the error installing libglpk-dev (got the idea from https://stackoverflow.com/a/71610023/1873521).

On Ubuntu 22.04: 
`sudo apt install libglpk-dev`

Just wanted to leave it here in case it is useful to someone.

**Comment by wlandau** [comment, marked as answer]
Definitely an `igraph` issue. I have been grappling with related variants of this issue on and off since I started `drake` in late 2016. I write GitHub Actions workflows for `targets` with [`sudo apt-get install libglpk-dev libglpk40`](https://github.com/ropensci/targets/blob/d17889faa45455cc3dab67d9de782809ff1d003c/.github/workflows/check.yaml#L44).

**Comment by szhorvat** [comment]
I would not call this an "igraph issue", rather GLPK is a dependency of igraph. Since CRAN does not provide binaries for Linux, you must compile R packages yourself, and you need to ensure that all dependencies are present, both at and after installation time.

Normally, if GLPK is not present when installing igraph, then GLPK support will simply be disabled. There will be no error messages.

If you see the error `libglpk.so.40: cannot open shared object file`, then one of the following must have happened:

 - At the time when you installed igraph, you did have GLPK installed as well. However, later you removed GLPK which broke igraph.
 - At the time when you installed igraph, a non-system version of GLPK was present in your path. Did you by any chance have an Anaconda environment active? If so, be sure to deactivate all conda environments before installing either igraph, or any other software. Otherwise, some libraries may be picked up from the conda environment. These may change, or may be incompatible with the system libraries on your machine, and either of those can either break igraph or cause misbehaviour.

For future reference, the Installation Troubleshooting FAQ is here: https://github.com/igraph/rigraph/wiki/Installation-Troubleshooting

**Reply by wlandau** [reply]
~~If igraph normally loads fine without glpk, then I wonder what is happening in GitHub Actions Linux instances that would cause trouble loading igraph unless there is an explicit apt install line for glpk. I am not at a computer to reproduce this, but it should be straightforward using usethis::use_github_action(“…”) plus igraph-dependent commands.~~ Never mind, https://github.com/wlandau/igraph-glpk-reprex/actions/runs/3221318429/jobs/5269105860 succeeded. Maybe the `igraph` installation in the cache was somehow out of sync with glpk.

**Comment by szhorvat** [comment]
@gorkang Since you can reproduce this problem, if you could answer the two questions [here](https://github.com/igraph/rigraph/issues/209#issuecomment-1273400011), that would be quite helpful for improving the igraph installation experience.

**Reply by gorkang** [reply]
@szhorvat  Not sure if I know how to respond to the two questions in a useful way:

1) Did you by any chance uninstall GLPK after installing igraph?: Not that I know of, but I have no idea how to check this

Did you have a conda environment active when you installed igraph? If by active you mean that I was using it in the same session, the answer is no. 

If you can guide me about how to better respond to the questions, I would be glad to help out.

**Reply by szhorvat** [reply]
Thanks so much for the responses! I'm hoping to be able to finally track down and resolve this.

What would be useful as a next step is to uninstall `igraph`, uninstall `libglpk-dev`, then restart R and try to install `igraph` again, the same way I asked someone else [here](https://github.com/igraph/rigraph/issues/209#issuecomment-1273525739). I know this is quite a bit of trouble, but if you could let me know whether installing igraph succeeds _without_ having `libglpk-dev` installed, that would be quite useful!

**Reply by gorkang** [reply]
I uninstalled igraph, uninstalled libglpk-dev and reinstalled igraph 1.3.5 without issues (Ubuntu 22.04).
